tests/driver_epd_bw_spi_disp_dev/main.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2019 Silke Hofstra
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup tests
+ * @{
+ *
+ * @file
+ * @brief       Test application for the black/white e-paper SPI display driver.
+ *
+ * @author      Silke Hofstra <silke@slxh.eu>
+ * @}
+ */
+
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "xtimer.h"
+
+#include "disp_dev.h"
+
+#include "epd_bw_spi.h"
+#include "epd_bw_spi_disp_dev.h"
+#include "epd_bw_spi_params.h"
+
+#include "pictures.h"
+
+const uint16_t riot_32_height = sizeof(riot_logo_32) / sizeof(riot_logo_32[0]);
+
+int main(void)
+{
+    epd_bw_spi_t epd;
+    int init = epd_bw_spi_init(&epd, epd_bw_spi_params, epd_bw_spi_controller,
+                               EPD_BW_SPI_DISPLAY_X, EPD_BW_SPI_DISPLAY_Y,
+                               EPD_BW_SPI_Y_INC_X_INC);
+
+    if (init != 0) {
+        printf("EPD_BW_SPI INIT FAILED: %i\n", init);
+        return 1;
+    }
+
+    disp_dev_t *dev = (disp_dev_t *)&epd;
+
+    dev->driver = &epd_bw_spi_disp_dev_driver;
+
+    xtimer_ticks32_t last_wakeup = xtimer_now();
+
+    while (1) {
+        /* Use full refreshes for the full-size logo */
+        epd_bw_spi_set_partial(dev, false);
```

Comment:
Exactly what I said above: do you expect people to call driver specific functions in an application that is using a generic API ?
But now that I see how you would like to use it, better check for the bounds used with the disp_dev `map` function and call the `epd_bw_spi_update_part`/`epd_bw_spi_update_full` accordingly.

Reply:
Yes, because the `disp_dev` API doesn't cover this fundamental property of e-paper displays.

> better check for the bounds used with the disp_dev map function and call the epd_bw_spi_update_part/epd_bw_spi_update_full accordingly.

The bounds don't necessarily have anything to do with wanting to do a full/partial refresh. The main reason to do a full refresh is to resolve the ghosting, otherwise a partial refresh will always give a better experience (and lower power draw).

Reply:
> The main reason to do a full refresh is to resolve the ghosting, otherwise a partial refresh will always give a better experience (and lower power draw).

Ok, I see. What about we use partial refresh all the time with disp_dev for now (and until we find a better solution with disp_dev) ? Extending the disp_dev interface for this use case is a possibility.

Reply:
What do you think about [the suggestion here](https://github.com/RIOT-OS/RIOT/pull/12509#discussion_r635155942)?
> Maybe instead of a `refresh_partial` flag it should be a `refresh_auto` flag that uses an internal counter to perform a full refresh once every n refreshes?